File: scripts/run_migration.py
import psycopg2

conn = psycopg2.connect(
    host='db.agoismqarzchkszihysr.supabase.co',
    port=5432,
    database='postgres',
    user='postgres',
    password='Zg233635'
)
conn.autocommit = True
cur = conn.cursor()

print('Connected to Supabase')

# Part 1: Tables
sql1 = """
CREATE TABLE IF NOT EXISTS system_config (
  key TEXT PRIMARY KEY,
  value TEXT NOT NULL,
  value_type TEXT DEFAULT 'string',
  description TEXT,
  updated_at TIMESTAMPTZ DEFAULT NOW()
);

INSERT INTO system_config (key, value, value_type, description) VALUES
('escrow_amount', '1', 'number', 'escrow points'),
('escrow_days', '30', 'number', 'escrow days'),
('min_inherits_for_release', '1', 'number', 'min inherits'),
('min_rating_for_release', '3.5', 'number', 'min rating'),
('release_on_first_inherit', 'true', 'boolean', 'release on first inherit')
ON CONFLICT (key) DO NOTHING;

CREATE TABLE IF NOT EXISTS transaction_ledger (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  from_agent_id UUID,
  to_agent_id UUID,
  amount DECIMAL(10,2) NOT NULL,
  transaction_type TEXT NOT NULL,
  reference_id UUID,
  description TEXT,
  created_at TIMESTAMPTZ DEFAULT NOW()
);

CREATE TABLE IF NOT EXISTS skill_inheritance_log (
  id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  skill_id UUID NOT NULL,
  inheritor_id UUID,
  points_spent DECIMAL(10,2) DEFAULT 1,
  task_success BOOLEAN DEFAULT true,
  created_at TIMESTAMPTZ DEFAULT NOW()
);

CREATE TABLE IF NOT EXISTS skill_reviews (
  review_id UUID PRIMARY KEY DEFAULT gen_random_uuid(),
  skill_id UUID NOT NULL,
  reviewer_id UUID,
  rating INTEGER NOT NULL CHECK (rating >= 1 AND rating <= 5),
  comment TEXT,
  created_at TIMESTAMPTZ DEFAULT NOW(),
  CONSTRAINT unique_review UNIQUE (skill_id, reviewer_id)
);

ALTER TABLE skill_escrow ADD COLUMN IF NOT EXISTS created_at TIMESTAMPTZ DEFAULT NOW();
ALTER TABLE skill_escrow ADD COLUMN IF NOT EXISTS expires_at TIMESTAMPTZ;
ALTER TABLE skill_escrow ADD COLUMN IF NOT EXISTS release_reason TEXT;
ALTER TABLE skill_escrow ADD COLUMN IF NOT EXISTS inherits_count INTEGER DEFAULT 0;
ALTER TABLE skill_escrow ADD COLUMN IF NOT EXISTS avg_rating DECIMAL(3,2) DEFAULT 0;
"""

try:
    cur.execute(sql1)
    print('Tables created successfully')
except Exception as e:
    print(f'Tables error: {e}')

# Part 2: Trigger
sql2 = """
CREATE OR REPLACE FUNCTION update_escrow_on_inherit()
RETURNS TRIGGER AS $$
BEGIN
  UPDATE skill_escrow
  SET inherits_count = inherits_count + 1
  WHERE skill_id = NEW.skill_id AND status = 'held';

  IF (SELECT value FROM system_config WHERE key = 'release_on_first_inherit') = 'true' THEN
    UPDATE skill_escrow
    SET status = 'released', release_reason = 'first_inherit', released_at = NOW()
    WHERE skill_id = NEW.skill_id AND status = 'held' AND inherits_count = 1;

    INSERT INTO transaction_ledger (from_agent_id, to_agent_id, amount, transaction_type, reference_id, description)
    SELECT NULL, se.owner_id, se.escrow_points, 'release', NEW.skill_id, 'First inheritance release'
    FROM skill_escrow se WHERE se.skill_id = NEW.skill_id AND se.release_reason = 'first_inherit';
  END IF;

  RETURN NEW;
END;
$$ LANGUAGE plpgsql;

DROP TRIGGER IF EXISTS trigger_escrow_inherit ON skill_inheritance_log;
CREATE TRIGGER trigger_escrow_inherit
AFTER INSERT ON skill_inheritance_log
FOR EACH ROW EXECUTE FUNCTION update_escrow_on_inherit();
"""

try:
    cur.execute(sql2)
    print('Trigger created successfully')
except Exception as e:
    print(f'Trigger error: {e}')

# Verify
cur.execute("SELECT key, value FROM system_config")
print('\n=== system_config ===')
for row in cur.fetchall():
    print(f'  {row[0]}: {row[1]}')

cur.execute("SELECT COUNT(*) FROM skill_escrow")
print(f'\nskill_escrow count: {cur.fetchone()[0]}')

cur.close()
conn.close()
print('\nMigration complete!')

File: scripts/run-migration.js
// 执行 migration 到 Supabase
// 运行方式: node scripts/run-migration.js

const { createClient } = require('@supabase/supabase-js')
const fs = require('fs')
const path = require('path')

const supabaseUrl = 'https://agoismqarzchkszihysr.supabase.co'
const supabaseServiceKey = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImFnb2lzbXFhcnpjaGtzemloeXNyIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc3NDE2NjgzMiwiZXhwIjoyMDg5NzQyODMyfQ.PliscqyQOXZsVby9p6aEOlCCWlGDRWzhauQ9PkQpjpE'

async function runMigration() {
  const supabase = createClient(supabaseUrl, supabaseServiceKey, {
    auth: { autoRefreshToken: false, persistSession: false }
  })

  // 读取 migration SQL
  const migrationPath = path.join(__dirname, '../supabase/migrations/escrow-config.sql')
  const sql = fs.readFileSync(migrationPath, 'utf8')

  console.log('Migration SQL length:', sql.length)
  
  // 分割 SQL 为多个语句执行
  // Supabase 不支持直接执行多语句，需要逐个执行
  
  // 简化：只创建必要的表和配置
  console.log('Creating system_config table...')
  
  // 用 RPC 执行 SQL（如果有的话）
  // 或者手动创建表
  
  try {
    // 先尝试创建 system_config 表
    const { error: configError } = await supabase.rpc('exec_sql', { query: `
      CREATE TABLE IF NOT EXISTS system_config (
        key TEXT PRIMARY KEY,
        value TEXT NOT NULL,
        value_type TEXT DEFAULT 'string',
        description TEXT,
        updated_at TIMESTAMPTZ DEFAULT NOW()
      );
    ` })
    
    if (configError) {
      console.log('RPC exec_sql not available, using table operations...')
      
      // 尝试直接插入配置（假设表已存在）
      const configs = [
        { key: 'escrow_amount', value: '1', value_type: 'number', description: '质押点数' },
        { key: 'escrow_days', value: '30', value_type: 'number', description: '质押周期（天）' },
        { key: 'min_inherits_for_release', value: '1', value_type: 'number', description: '最少继承次数' },
        { key: 'min_rating_for_release', value: '3.5', value_type: 'number', description: '最低好评率' },
        { key: 'release_on_first_inherit', value: 'true', value_type: 'boolean', description: '是否首传即退' }
      ]
      
      // 尝试创建表（用 Supabase 表 API）
      // 实际上 Supabase REST API 不能创建表，需要用 SQL Editor
      
      console.log('Please run the migration SQL in Supabase SQL Editor:')
      console.log('---')
      console.log(sql.substring(0, 500) + '...')
      console.log('---')
      
      // 保存 SQL 到临时文件方便复制
      const tempPath = path.join(__dirname, '../temp-migration.sql')
      fs.writeFileSync(tempPath, sql)
      console.log('SQL saved to:', tempPath)
      
      return
    }
    
    console.log('system_config table created')
    
  } catch (err) {
    console.error('Migration error:', err.message)
    console.log('Please run the migration SQL manually in Supabase SQL Editor')
    console.log('File location: supabase/migrations/escrow-config.sql')
  }
}

runMigration()